internal/p2p/channel.go

```diff
@@ -259,17 +264,85 @@ func (ch *libp2pChannelImpl) canonicalizedTopicName() string {
 }
 
 func (ch *libp2pChannelImpl) Receive(ctx context.Context) *ChannelIterator {
+	// TODO: consider caching an iterator in the channel, or
+	// erroring if this gets called more than once.
+	//
+	// While it's safe to register a handler more than once, we
+	// could get into a dodgy situation where if you call receive
+	// more  than once, the subsequently messages won't be routed
+	// correctly.
 	iter := &ChannelIterator{
 		pipe: make(chan Envelope),
 	}
 
+	ch.host.SetStreamHandler(protocol.ID(ch.canonicalizedTopicName()), func(stream network.Stream) {
+		buf := bufio.NewReader(stream)
+		// TODO: change the max delimited message into
+		// something that's passed in as a parmeter.
+		reader := protoio.NewDelimitedReader(buf, 1023)
+
+		remote := stream.Conn().RemotePeer()
+
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+
+			payload := proto.Clone(ch.chDesc.MessageType)
+
+			// TODO: context might expire while reading
+			// the message, and so we could end up
+			// blocking here without any way to clean
+			// things up.
+			if _, err := reader.ReadMsg(payload); err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+
+				}
+				// TODO: this should cause the channel to
+				// report an error, or abort the receiver.
+				_ = stream.Close()
+			}
+			select {
+			case <-ctx.Done():
+				return
+			case iter.pipe <- Envelope{
```

Comment:
I think I am missing something basic here—why are we pushing messages from the stream into the same channel as the results of the subscription below?

Reply:
the `iter.pipe` is the thing that consumers of the channel (e.g. the reactors) use to get messages, so we have to put messages that we got from the topic subscription (e.g. things that were broadcast) and that we receive directly from peers into the same channel and let the reactors sort things out. 

This is the temporary part of the whole mix. I think most of our existing point-to-point communication (e.g. evidence, votes, and mempool) will just become broadcasts into gossipsub, and the only point-to-point communication will be for statesync and blocksync, but I think we have to hoist everything off the old mechanisms and clean up the tests, so this is sort of a shim for the moment.